### PR TITLE
EAR-2449 Session Timeout

### PR DESCRIPTION
**Pull request by sanjeevz3009**
Implement session time-out for Author.

Credit to @martyncolmer and @farres1 for the assistance/ part-taking in discussions on the direction of this ticket.

### What is the context of this PR?

To add functionality for Author users to automatically have their sessions timed out and require the users to sign in once every 7 days.

Jira ticket: https://jira.ons.gov.uk/browse/EAR-2449

### Files changed: `src/App/MeContext.js`

Added the session timeout functionality inside the first `useEffect` within the `ContextProvider` function. 

### How to review

To test the functionality you can replace this line of code 
`const sessionDuration = 1000 * 60 * 60 * 24 * 7;` which is 7 days in milliseconds to 5 or 10 seconds by copying and pasting either this line
`sessionDuration = 10000` for 10 seconds (in milliseconds) or `sessionDuration = 5000` for 5 seconds (in milliseconds).

Next sign into the Author application and you can see it time out the user after 5 seconds or 10 seconds or after the duration set by the developer.

This should work regardless of whether the user is on the Author application or not. It should work even if the user closes the application and comes back.

### Important information: Things to consider before merging
All the users who are already signed in and have been signed in for 7 days will be signed out immediately when the code changes go into production. This will affect the majority of the users, as we are assuming the majority of the users have been signed in for more than 7 days and haven't signed out.

For more information please refer to the comment in the JIRA ticket: https://jira.ons.gov.uk/browse/EAR-2449

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author/src/App/MeContext.js [hunk 1]
```diff
@@ -89,10 +89,34 @@ const ContextProvider = ({ history, client, children }) => {
 
   useEffect(() => {
     // be aware that the return from auth.onAuthStateChanged will change on firebase ver 4.0
-    // https://firebase.google.com/docs/reference/js/v8/firebase.auth.Auth#onauthstatechanged
+    // https://firebase.google.com/docs/reference/js/v8/firebase.auth.Auth#onauthstatechanged]
```
Comment (Paul-Joel) [inline eq-author/src/App/MeContext.js:92]: Square bracket has been added to the end of this line. Can't see why.

File: eq-author/src/App/MeContext.js [hunk 2]
```diff
@@ -89,10 +89,34 @@ const ContextProvider = ({ history, client, children }) => {
...
+      // If the user is not authenticated, the session timeout is cleared using clearTimeout.
+      let sessionTimeout = null;
+      if (user === null || user === undefined) {
+        sessionTimeout && clearTimeout(sessionTimeout);
+        sessionTimeout = null;
+      } else {
+        // If the user is authenticated, it retrieves the ID token result and calculates the session duration.
+        user.getIdTokenResult().then((idTokenResult) => {
+          const authTime = idTokenResult.claims.auth_time * 1000;
+          // The session duration is set to 7 days in this example.
```
Comment (Paul-Joel) [inline eq-author/src/App/MeContext.js:109]: This isn't an example, it's the actual code.

File: eq-author/src/App/MeContext.js [hunk 3]
```diff
@@ -89,10 +89,34 @@ const ContextProvider = ({ history, client, children }) => {
...
+          const sessionDuration = 7 * 24 * 60 * 60 * 1000;
```
Comment (Paul-Joel) [inline eq-author/src/App/MeContext.js:110]: I'm not keen on this as a sum - it really ought to be just the result of the sum. Putting the calculation in a comment after is fine to show how the value was determined though.

**Review by Paul-Joel — CHANGES_REQUESTED** [review]
Couple of minor tweaks.

Testing this however raises a deployment issue - running the code immediately signs you out. This means that it would do that to every user as soon as it is deployed.
Suspect that the code cannot be merged until a specific rollout point is agreed, and all users are warned and notified.

Also, is there an implementation for Author v2? All changes in v1 must be either implemented in v2, or a ticket raised for v2 work to ensure that the changes are not missed. As v2 has already implemented sign in, it really ought to be implemented in v2 now.

File: eq-author/src/App/MeContext.js (hunk 2 above)
Comment (sanjeevz3009) [reply eq-author/src/App/MeContext.js]: Comment updated!

File: eq-author/src/App/MeContext.js (hunk 3 above)
Comment (sanjeevz3009) [reply eq-author/src/App/MeContext.js]: The calculation for the session duration updated!

**Review by Paul-Joel — APPROVED** [review]

**Review by sudeepkunhis — APPROVED** [review]

**Pull request merged**